### PR TITLE
Ea/42

### PR DESCRIPTION
**Pull request by eamouzou**
- User stories 42 - 48
- Create tests, routes, controllers(actions), and views for admin merchant activation, add, delete and edit functionality.
- Revise relevant models

### CODE REVIEW

File: spec/models/item_spec.rb [hunk 1]
```diff
@@ -56,6 +55,14 @@
...
       expect(@chain.no_orders?).to eq(false)
     end
+
+    it "updates_active" do
+      expect(@chain.active?).to eq(true)
+      @chain.updates_active
+      expect(@chain.active?).to eq(false)
+      @chain.updates_active
+      expect(@chain.active?).to eq(true)
+    end
```
Comment (eamouzou) [inline spec/models/item_spec.rb:65]: toggles the active status for item model

File: spec/features/items/new_spec.rb [hunk 2]
```diff
@@ -70,7 +70,7 @@
 
       click_button "Create Item"
 
-      expect(page).to have_content("Name can't be blank and Inventory can't be blank")
+      expect(page).to have_content("Name can't be blank, Inventory can't be blank, and Inventory is not a number")
```
Comment (eamouzou) [inline spec/features/items/new_spec.rb:73]: revised for added validation of inventory numericality

File: spec/features/items/edit_spec.rb [hunk 3]
```diff
@@ -65,7 +65,7 @@
 
         click_button "Update Item"
 
-        expect(page).to have_content("Name can't be blank and Image can't be blank")
+        expect(page).to have_content("Name can't be blank")
```
Comment (eamouzou) [inline spec/features/items/edit_spec.rb:68]: revised because image cannot be required

File: spec/features/admins/items/index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+    end
+
+    expect(current_path).to eq("/admin/merchants/#{@mike.id}/items/#{@item2.id}/edit")
+    fill_in :name, with: "Hornswoggle"
+    fill_in :inventory, with: "-1"
+
+    click_button('Update Item')
+    expect(page).to have_content("Price is not a number and Inventory must be greater than 0")
+  end
+
```
Comment (eamouzou) [inline spec/features/admins/items/index_spec.rb:185]: add test for admin merchants add, delete, and edit functionality on index page

File: db/schema.rb [hunk 5]
```diff
@@ -31,12 +31,12 @@
     t.string "name"
     t.string "description"
     t.integer "price"
-    t.string "image"
     t.boolean "active?", default: true
     t.integer "inventory"
     t.bigint "merchant_id"
     t.datetime "created_at", null: false
     t.datetime "updated_at", null: false
+    t.string "image", default: "https://i.picsum.photos/id/866/200/300.jpg"
```
Comment (eamouzou) [inline db/schema.rb:39]: provide default image

File: db/migrate/20200225031842_add_image_to_items.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+class AddImageToItems < ActiveRecord::Migration[5.1]
+  def change
+    add_column :items, :image, :string, default: "https://i.picsum.photos/id/866/200/300.jpg"
+    Item.all.where(image: nil).update_all(image: "https://i.picsum.photos/id/866/200/300.jpg")
+    Item.all.where(image: '').update_all(image: "https://i.picsum.photos/id/866/200/300.jpg")
+  end
```
Comment (eamouzou) [inline db/migrate/20200225031842_add_image_to_items.rb:6]: add default image

File: config/routes.rb [hunk 7]
```diff
@@ -70,6 +70,12 @@
     patch '/orders/:id', to: "dashboard#update"
     patch '/merchants/:id', to: "merchants#update"
     get '/merchants/:merchant_id', to: "dashboard#show"
+    get '/merchants/:merchant_id/items', to: "items#index"
+    patch '/merchants/:merchant_id/items/:item_id', to: "items#update"
+    delete '/merchants/:merchant_id/items/:item_id', to: "items#destroy"
+    get '/merchants/:merchant_id/items/new', to: "items#new"
+    post '/merchants/:merchant_id/items', to: 'items#create'
+    get '/merchants/:merchant_id/items/:item_id/edit', to: 'items#edit'
```
Comment (eamouzou) [inline config/routes.rb:78]: These are the routes for admin's ability to add an item, delete an item, change the activation status, and edit an item.

File: app/views/admin/items/new.html.erb [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+
+    <%= label_tag :image %>
+    <%= text_field_tag :image %>
+
+    <%= label_tag :inventory %>
+    <%= text_field_tag :inventory %>
+
+    <%= submit_tag 'Add Item' %>
+  <% end %>
+</center>
```
Comment (eamouzou) [inline app/views/admin/items/new.html.erb:22]: form to add an item

File: app/views/admin/items/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    <p>Description: <%= item.description %> </p>
+    <p>Price: <%=number_to_currency(item.price) %> </p>
+    <%= image_tag(item.image, height: '100', width: '100') %>
+    <% if item.active? %>
+      <p>Status: Active &nbsp; &nbsp; &nbsp;
+      <%= link_to "Deactivate #{item.name}", "/admin/merchants/#{@merchant.id}/items/#{item.id}", method: :patch %></p>
+    <% else %>
+      <p>Status: Inactive &nbsp; &nbsp; &nbsp;
+      <%= link_to "Activate #{item.name}", "/admin/merchants/#{@merchant.id}/items/#{item.id}", method: :patch %></p>
+    <% end %>
```
Comment (eamouzou) [inline app/views/admin/items/index.html.erb:16]: shows relevant link based on item's active status

File: app/views/admin/items/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+    <p>Inventory: <%= item.inventory %> </p>
+    <%= link_to "Edit #{item.name}", "/admin/merchants/#{@merchant.id}/items/#{item.id}/edit", method: :get %> &nbsp; &nbsp; &nbsp;
+    <% if item.no_orders? == true  %>
+      <%= link_to "Delete #{item.name}", "/admin/merchants/#{@merchant.id}/items/#{item.id}", method: :delete %>
+    <% end %>
```
Comment (eamouzou) [inline app/views/admin/items/index.html.erb:21]: You can only delete if there are no orders.

File: app/views/admin/items/edit.html.erb [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+
+    <%= label_tag :image %>
+    <%= text_field_tag :image, "#{@item.image}" %>
+
+    <%= label_tag :inventory %>
+    <%= text_field_tag :inventory, "#{@item.inventory}" %>
+
+    <%= submit_tag 'Update Item' %>
+  <% end %>
+</center>
```
Comment (eamouzou) [inline app/views/admin/items/edit.html.erb:22]: form to edit item with pre-populated fields

File: app/models/item.rb [hunk 12]
```diff
@@ -40,4 +39,12 @@ def self.least_popular
...
     .limit(5).to_a
   end
+
+  def updates_active
+    if active? == true
+      update(active?: false)
+    else
+      update(active?: true)
+    end
+  end
```
Comment (eamouzou) [inline app/models/item.rb:49]: toggles the active status for item model

File: app/controllers/admin/items_controller.rb [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+  end
+
+  def update
+    @item = Item.find(params[:item_id])
+    if params[:commit] == "Update Item"
+      @item.update(item_params)
+      save_process(@item, render_arg = :edit, message = "#{@item.name} is updated.")
+    else
+      activation_process(@item)
+    end
```
Comment (eamouzou) [inline app/controllers/admin/items_controller.rb:13]: Update is the action used to toggle the active status and to update an item's attributes. The conditional filters which process based on the presence of the Update Item commit param.

File: app/controllers/admin/items_controller.rb [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+  def save_process(item, render_arg, message)
+    default_image(item)
+    if item.save
+      flash[:saved] = message
+      redirect_to "/admin/merchants/#{params[:merchant_id]}/items"
+    else
+      flash[:error] = item.errors.full_messages.to_sentence
+      render render_arg
+    end
+  end
```
Comment (eamouzou) [inline app/controllers/admin/items_controller.rb:68]: This is used in update and create with the appropriate arguments passed in. It assigns a default image if the field is empty or nil. It flashes the appropriate message and redirects or renders based on completeness.

File: app/controllers/admin/items_controller.rb [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+  def activation_process(item)
+    item.updates_active
+    if item.active? == false
+      flash[:update] = "#{item.name} is no longer for sale."
+    else
+      flash[:update] = "#{item.name} is now available for sale."
+    end
+    redirect_to "/admin/merchants/#{params[:merchant_id]}/items"
```
Comment (eamouzou) [inline app/controllers/admin/items_controller.rb:56]: changes active status from active to inactive or vice-versa and flashes the relevant message.

**Review by agallant121 — APPROVED** [review]

**Pull request merged**